### PR TITLE
Fix a crash in X509v3_asid_subset()

### PR DESCRIPTION
**Pull request by mattcaswell**
If the asnum or rdi fields are NULL and the ASIdentifiers are otherwise
subsets then this will result in a crash. Of note is that rdi will usually
be NULL.

Reported by Theo Buehler (@botovq)

### CODE REVIEW

File: crypto/x509/v3_asid.c
Comment (botovq) [inline crypto/x509/v3_asid.c:703]: I think requiring  `&& b->asnum == NULL` here is too stringent. If `a` only specifies RDI resources, but no AS numbers, it is still a subset of `b` even if `b` happens to specify resources of both type.

The analogous check in `X509v3_addr_subset()` considers an `a` only specifying IPv4 addresses a subset of a `b` specifying IPv4 and IPv6 addresses.

File: crypto/x509/v3_asid.c
Comment (botovq) [inline crypto/x509/v3_asid.c:711]: Same here for `&& b->rdi == NULL`.

**Review by botovq — COMMENTED** [review]
The rest looks good.

File: crypto/x509/v3_asid.c
Comment (mattcaswell) [reply crypto/x509/v3_asid.c]: Good point. Fixed.

**Comment by mattcaswell** [comment]
Updated with fixes following feedback.

**Review by botovq — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
Ping for second review @openssl/committers

**Review by InfoHunter — APPROVED** [review]

**Comment by tmshort** [comment]
Merged to `master`, `openssl-3.0` and `OpenSSL_1_1_1-stable`. Thank you!

**Pull request closed without merging**
